### PR TITLE
Finality loop: get block justification and authorities change by consensus engine ID

### PR DESCRIPTION
**Pull request by serban300**

### CODE REVIEW

File: primitives/header-chain/src/lib.rs [hunk 1]
```diff
@@ -77,18 +77,35 @@ pub trait FinalityProof<Number>: Clone + Send + Sync + Debug {
 	fn target_header_number(&self) -> Number;
 }
 
-/// Find header digest that schedules next GRANDPA authorities set.
-pub fn find_grandpa_authorities_scheduled_change<H: HeaderT>(
-	header: &H,
-) -> Option<sp_finality_grandpa::ScheduledChange<H::Number>> {
-	let id = OpaqueDigestItemId::Consensus(&GRANDPA_ENGINE_ID);
+pub trait ConsensusLogReader {
```
Comment (svyatonik) [inline primitives/header-chain/src/lib.rs:80]: Can you, please, leave some docs for this public trait? Thanks!

File: primitives/header-chain/src/lib.rs [hunk 2]
```diff
@@ -77,18 +77,35 @@ pub trait FinalityProof<Number>: Clone + Send + Sync + Debug {
...
+	type ConsensusLog;
 
-	let filter_log = |log: ConsensusLog<H::Number>| match log {
-		ConsensusLog::ScheduledChange(change) => Some(change),
-		_ => None,
-	};
+	fn schedules_authorities_change(digest: &Digest) -> bool;
+}
+
+pub struct GrandpaConsensusLogReader<Number>(sp_std::marker::PhantomData<Number>);
```
Comment (svyatonik) [inline primitives/header-chain/src/lib.rs:87]: Some docs. please. Thanks!

File: relays/lib-substrate-relay/src/finality/engine.rs [hunk 3]
```diff
@@ -183,11 +188,14 @@ impl<C: ChainWithGrandpa> Engine<C> for Grandpa<C> {
 		// If initial header changes the GRANDPA authorities set, then we need previous authorities
 		// to verify justification.
 		let mut authorities_for_verification = initial_authorities_set.clone();
-		let scheduled_change = find_grandpa_authorities_scheduled_change(&initial_header);
+		let scheduled_change =
+			GrandpaConsensusLogReader::<<C::Header as Header>::Number>::find_authorities_change(
```
Comment (svyatonik) [inline relays/lib-substrate-relay/src/finality/engine.rs:192]: Probably use `BlockNumberOf<C>` here?

File: primitives/header-chain/src/lib.rs [hunk 4]
```diff
@@ -77,18 +77,35 @@ pub trait FinalityProof<Number>: Clone + Send + Sync + Debug {
 	fn target_header_number(&self) -> Number;
 }
 
-/// Find header digest that schedules next GRANDPA authorities set.
-pub fn find_grandpa_authorities_scheduled_change<H: HeaderT>(
-	header: &H,
-) -> Option<sp_finality_grandpa::ScheduledChange<H::Number>> {
-	let id = OpaqueDigestItemId::Consensus(&GRANDPA_ENGINE_ID);
+pub trait ConsensusLogReader {
+	const ID: ConsensusEngineId;
```
Comment (svyatonik) [inline primitives/header-chain/src/lib.rs:81]: Is this const used (or planned to be) used anywhere? I wonder why do we have similar const in both `Engine` trait and here?

Comment (svyatonik) [reply primitives/header-chain/src/lib.rs:81]: Same question for `type ConsensusLog` - is it used anywhere?

**Review by svyatonik — APPROVED** [review]
lgtm, left couple of questions

File: primitives/header-chain/src/lib.rs (hunk 4 above)
Comment (serban300) [reply primitives/header-chain/src/lib.rs]: I removed them. They are not used.

**Review by svyatonik — APPROVED** [review]
Thank you!

**Pull request merged**